Debugger: Get target memory in load/store instructions

TryTwo opened the pull request:
Backend:
Updates disassembler instructions. Mostly spaces and commas. PS load/store uses 0x notation. Required for new features.
Adds function to PPCDebugInterface to calculate target memory address for load/store instructions.

Code widget new feature:
-Context menu for copying or showing memory target in load/store instructions. Only works if the PC is near the instruction (register consistency).

/edit removed trace and diff for their own PRs
/edit2 changed breakpoint memory to 'show in memory'.

TryTwo commented:
I could follow the "show in memory" context item to send the target memory address to the memory widget, rather than breakpoint or even copy it. If that's better?

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+  u32 roffs = 0;
+  std::regex re(",\\W*(-?0[xX]?[0-9a-fA-F]*|r\\d+)[^r^s]*.(p|toc|\\d+)");
+  std::smatch match;
+  std::regex_search(instruction, match, re);
```
The return value of this should probably be checked.

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
 
+u32 PPCDebugInterface::GetMemoryAddressFromInstruction(std::string instruction)
+{
+  int offs = 0;
```
```suggestion
  u32 offset = 0;
```
(and handle necessary conversions to u32 explicitly). Otherwise expressions further down are intermixing signed and unsigned types implicitly.

TryTwo replied:
offset might be negative (-0x6000 rtoc). roffs is always a memory address.  Do I use long long for everything and convert to u32 at the end? Or is there a trick I'm missing.

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+
+  // Output: match.str(1): either -+0xNNNN, 0, or rNN. Check for r later to see if a gpr needs to be
+  // loaded. match.str(2): will either be p, toc, or NN. Always a gpr.
+  if (match.str(1).find("r") != std::string::npos)
```
The result of `match.str(1)` should be stored to its own variable, otherwise this is redundantly creating the same std::string instance inside the if statements. `"r"` can also be turned into a character literal to allow library implementations to potentially take a cheaper find algorithm (it's quicker to find a single character than to find a string).

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+  // loaded. match.str(2): will either be p, toc, or NN. Always a gpr.
+  if (match.str(1).find("r") != std::string::npos)
+  {
+    const int i = stoi(match.str(1).erase(0, 1), nullptr, 10);
```
```suggestion
    const int register_index = std::stoi(match.str(1).erase(0, 1), nullptr, 10);
```

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+
+  // sp and rtoc need to be converted to 1 and 2. First letter is already removed.
+  int i;
+  if (match.str(2).find("p") != std::string::npos)
```
Ditto about storing the results of `match.str(2)` into a const variable. `"p"` can also be changed into a character literal.

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+  else
+    i = std::stoi(match.str(2), nullptr, 10);
+
+  u32 addr = GPR(i);
```
```cpp
return GPR(i) + offset + register_offset;
```

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+u32 PPCDebugInterface::GetMemoryAddressFromInstruction(std::string instruction)
+{
+  int offs = 0;
+  u32 roffs = 0;
```
```suggestion
  u32 register_offset = 0;
```

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.h:
```diff
@@ -76,6 +76,7 @@ class PPCDebugInterface final : public Common::DebugInterface
   void RunToBreakpoint() override;
   int GetColor(unsigned int address) override;
   std::string GetDescription(unsigned int address) override;
+  u32 GetMemoryAddressFromInstruction(std::string instruction);
```
This can likely be a const qualified member function.

lioncash replied:
I had mean that the function can be const qualified, like:

```cpp
u32 GetMemoryAddressFromInstruction(std::string instruction) const;
```

not like:

```cpp
const u32 GetMemoryAddressFromInstruction(std::string instruction);
```
(which is qualifying the return value, not the function)

lioncash commented on Source/Core/Common/GekkoDisassembler.h:
```diff
@@ -128,7 +128,25 @@ class GekkoDisassembler final
   }
 
-  static int SEX12(u32 x) { return x & 0x800 ? (x | 0xFFFFF000) : x; }
+  static std::string psq_offs(u32 val)
```
This should be put in the cpp file, since it's only used there.

lioncash commented on Source/Core/Common/GekkoDisassembler.h:
```diff
@@ -128,7 +128,25 @@ class GekkoDisassembler final
-  static int SEX12(u32 x) { return x & 0x800 ? (x | 0xFFFFF000) : x; }
+  static std::string psq_offs(u32 val)
+  {
+    if (val == 0)
```
```cpp
if (val == 0)
  return "0";

if ((val & 0x800) != 0)
  return StringFromFormat("-0x%.4X", (~val & 0xfff) + 1);

return StringFromFormat("0x%.4X", val);
```

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -208,6 +209,55 @@ void CodeViewWidget::ReplaceAddress(u32 address, ReplaceWith replace)
 
+bool CodeViewWidget::IsInstructionLoadStore(std::string instruction)
+{
+  // Could add check for context address being near PC, because we need gprs to be correct for the
```
```cpp
return (instruction.compare(0, 2, "st") == 0 ||
        instruction.compare(0, 1, "l") == 0 ||
        instruction.compare(0, 5, "psq_l") == 0 ||
        instruction.compare(0, 5, "psq_s") == 0) &&
        instruction.compare(0, 2, "li") != 0);
```

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -208,6 +209,55 @@ void CodeViewWidget::ReplaceAddress(u32 address, ReplaceWith replace)
   Update();
 }
 
+bool CodeViewWidget::IsInstructionLoadStore(std::string instruction)
```
This can just be a local function, since it doesn't depend on any class state.

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -208,6 +209,55 @@ void CodeViewWidget::ReplaceAddress(u32 address, ReplaceWith replace)
+
+void CodeViewWidget::OnCopyTargetAddress()
+{
+  const std::string codeline = PowerPC::debug_interface.Disassemble(GetContextAddress());
```
```suggestion
  const std::string code_line = PowerPC::debug_interface.Disassemble(GetContextAddress());
```

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -208,6 +209,55 @@ void CodeViewWidget::ReplaceAddress(u32 address, ReplaceWith replace)
+  if (!IsInstructionLoadStore(codeline))
+    return;
+
+  const u32 addr = PowerPC::debug_interface.GetMemoryAddressFromInstruction(codeline);
```
```suggestion
  const u32 address = PowerPC::debug_interface.GetMemoryAddressFromInstruction(codeline);
```

You don't need to mince words, we aren't running on an interpreter on a system with so little memory that shortening names is needed to save memory ;)

TryTwo replied:
I don't mind making it longer, but all the other functions use addr instead of address, so I probably should leave it for consistency?

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -208,6 +209,55 @@ void CodeViewWidget::ReplaceAddress(u32 address, ReplaceWith replace)
+
+void CodeViewWidget::OnBreakpointTargetAddress()
+{
+  const std::string codeline = PowerPC::debug_interface.Disassemble(GetContextAddress());
```
```suggestion
  const std::string code_line = PowerPC::debug_interface.Disassemble(GetContextAddress());
```

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -208,6 +209,55 @@ void CodeViewWidget::ReplaceAddress(u32 address, ReplaceWith replace)
+  if (!IsInstructionLoadStore(codeline))
+    return;
+
+  const u32 addr = PowerPC::debug_interface.GetMemoryAddressFromInstruction(codeline);
```
```suggestion
  const u32 address = PowerPC::debug_interface.GetMemoryAddressFromInstruction(codeline);
```

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,40 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
   return g_symbolDB.GetDescription(address);
 }
 
+u32 PPCDebugInterface::GetMemoryAddressFromInstruction(std::string instruction)
```
It may be preferable to use a `std::optional<u32>` for this. To handle the mentioned error case relating to checking the return value of `std::regex_search`.

TryTwo replied:
This is a tough one. For Trace I always need this to reliably return a memory address. I filter what can feed into this, to make sure it's a load/store, and so it should reliably work. If I could think of an instance of it failing, I could decide on how handle it. I'm thinking maybe just return 0 for false.

TryTwo commented:
Thanks for the detailed review, I'll work on it.  For tracing and differencing, I included it because it relies on the disassembler and PPCdebuginfo changes.  If I split the PR do I put those files in each or a note that says one PR relies on another to work?

Should I split trace and diff into different PRs? I kept them together because they are each in their own cpp, but that might be poor logic. I could also just break off the diff changes, which don't rely on the dissassembler and PPC changes.

TryTwo commented:
I found one major issue with GekkoDisassembler. On load/stores that use a register for the offset (lwzx) if the offset register value is 0, then the offset is 0 and NOT r0. All others cases are normal (offset = 1 -> r1).  I don't know how to correctly fix that.  I will edit my PPCDebugInterface to deal with the error instead.

This error also applies to things like icbi. icbi r0, r4 should be icbi 0, r4.

TryTwo commented:
Fixed offset = 0 instead of r0. 
Removed breakpoint target memory address contexty item.
Added show target memory in the memory widget context item. -- In practice you often want to BP it on write or on read, not both, so sending it to the memory widget and BPing it there works better..

lioncash commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -6,6 +6,7 @@
 
 #include <algorithm>
 #include <cmath>
+#include <regex>
```
This doesn't appear to actually be used within the file anywhere.

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,51 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+  }
+  else
+  {
+    offset = (u32)std::stoi(offset_match, nullptr, 16);
```
static_cast

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,51 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
 
+const u32 PPCDebugInterface::GetMemoryAddressFromInstruction(std::string instruction)
+{
+  u32 offset = 0;
```
All of these variables can be moved closer to their actual usage areas.

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,51 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+{
+  u32 offset = 0;
+  u32 register_offset = 0;
+  char is_reg = 'r';
```
These char variables can be marked `constexpr`

lioncash commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -324,6 +325,51 @@ std::string PPCDebugInterface::GetDescription(unsigned int address)
+  else
+    i = std::stoi(register_match, nullptr, 10);
+
+  u32 base_address = GPR(i);
```
```suggestion
  const u32 base_address = GPR(i);
```

TryTwo commented:
Fixed everything I think needs fixing. Any open issues are left open for further comment, but may not need anything done.

Also tweaked: 
PPCDebugInfo: I saw after recent changes I could remove one of the offset variables and just use one variable for both.
CodeViewWidget: Slightly reordered my additions in the context men, and "Show in memory" -> "Show instruction in &memory", since we have two show in memory options now I thought it should be more verbose.

Pokechu22 commented:
@dolphin-emu-bot rebuild

Pokechu22 commented:
@dolphin-emu-bot rebuild

TryTwo commented:
Would still be nice to see this old PR get accepted.

This PR has important base changes for the use of regex on PPC instructions. It is shared by my code tracing PR, which I haven't had time to work on.  This would be a small, but definitely step in that direction and opening up new useful features to be added.  After this PR, my plan is to implement the base tracing logic in another PR with a basic usable feature, then perhaps tackle the full code tracing PR again.

Pokechu22 commented:
@dolphin-emu-bot rebuild

sepalani commented on Source/Core/Common/DebugInterface.h:
```diff
@@ -72,6 +72,10 @@ class DebugInterface
   virtual void WriteExtraMemory(int /*memory*/, u32 /*value*/, u32 /*address*/) {}
   virtual u32 ReadExtraMemory(int /*memory*/, u32 /*address*/) const { return 0; }
   virtual u32 ReadInstruction(u32 /*address*/) const { return 0; }
+  virtual std::optional<u32> GetMemoryAddressFromInstruction(std::string /*instruction*/) const
```
Can't the parameter be passed as `const std::string&` instead?

sepalani commented on Source/Core/Common/GekkoDisassembler.cpp:
```diff
@@ -459,14 +466,11 @@ std::string GekkoDisassembler::fd_ra_rb(u32 in, int mask)
   if (mask)
   {
     if (mask & 4)
-      result += fmt::format("f{},", PPCGETD(in));
```
Feel free to correct me if I'm wrong as I don't know these instructions much.

Does your approach work when `mask & 4` is 0 and `mask & 2` is 1. Won't it produce a heading coma. If not, you should probably leave a comment stating why this approach is working.

Otherwise, what's wrong with the previous approach? If you're looking just to add a space character, can't you do it directly? Maybe using `fmt::join` would be cleaner.

TryTwo replied:
This confuses me, as it's only called by fdab and fdab only receives masks of 7 (all true). It could be some remnant of earlier code, but I wasn't sure I should change it completely. I can rewrite it and remove the mask if people are okay with it. Regardless, mask & 4 must always be true, as it's a float operation and that requires F{} register at the start.

TryTwo replied:
After thinking about it some more, there's no reason to keep the mask stuff. Removed the mask parameter from fdab instructions.

sepalani commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -375,6 +376,54 @@ std::string PPCDebugInterface::GetDescription(u32 address) const
   return g_symbolDB.GetDescription(address);
 }
 
+std::optional<u32> PPCDebugInterface::GetMemoryAddressFromInstruction(std::string instruction) const
```
`const std::string&`

sepalani commented on Source/Core/Core/Debugger/PPCDebugInterface.h:
```diff
@@ -76,6 +76,7 @@ class PPCDebugInterface final : public Common::DebugInterface
 
   u32 ReadExtraMemory(int memory, u32 address) const override;
   u32 ReadInstruction(u32 address) const override;
+  std::optional<u32> GetMemoryAddressFromInstruction(std::string instruction) const override;
```
`const std::string&`

sepalani commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -235,6 +235,15 @@ static bool IsBranchInstructionWithLink(std::string_view ins)
          StringEndsWith(ins, "la+") || StringEndsWith(ins, "l-") || StringEndsWith(ins, "la-");
 }
 
+bool IsInstructionLoadStore(std::string instruction)
```
`const std::string&`

sepalani commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -561,6 +573,12 @@ void CodeViewWidget::OnContextMenu()
 
+  for (auto* action : {copy_target_memory, show_target_memory})
+  {
+    action->setEnabled(Core::GetState() == Core::State::Paused &&
```
The expression is evaluated multiple times. This should be moved outside the for loop and put into a constant.

Something like this or similar:
```c++
const bool is_valid_target_addr = Core::GetState() == Core::State::Paused && IsInstructionLoadStore(PowerPC::debug_interface.Disassemble(addr))
for (auto* action : {copy_target_memory, show_target_memory})
{
  action->setEnabled(is_valid_target_addr);
}
```

TryTwo commented:
Side note: GekkoDisasm doesn't recognize when 0 needs to be 0 and not r0.  I deal with this bug in my regex logic. lfsx f1, r0, r4 should be lfs f1, 0, r4 for example. Too much work for me to identify every case and fix, I think. I'll check the bug reports to see if it's noted.

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
I think my commits are all good now.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani reviewed (changes requested):
> I think my commits are all good now, except for squashing. Am I expected to hit the re-request review every time I finish updating it? I don't know.

IIRC, it sends a notification to the reviewers. I think it's fine to use it when your PR is ready to be reviewed.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -574,11 +594,46 @@ void CodeViewWidget::OnCopyAddress()
+
+  if (addr)
+    QApplication::clipboard()->setText(
+        QStringLiteral("%1").arg(addr.value(), 8, 16, QLatin1Char('0')));
```
Good to know. I had thought value might be preferred for clarity.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeViewWidget.cpp:
```diff
@@ -235,6 +235,15 @@ static bool IsBranchInstructionWithLink(std::string_view ins)
+{
+  // Could add check for context address being near PC, because we need gprs to be correct for the
+  // load/store.
+  return ((instruction.compare(0, 2, "st") == 0 || instruction.compare(0, 1, "l") == 0 ||
```
I noticed the function above this one and thought I should rewrite mine to be similar to it.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented on Source/Core/Common/GekkoDisassembler.cpp:
```diff
@@ -189,12 +189,19 @@ static std::string ldst_offs(u32 val)
 {
-  if ((x & 0x800) != 0)
-    return static_cast<int>(x | 0xFFFFF000);
+  if (val == 0)
```
The `else` statement can be removed and the braces can be removed like this:
```c++
if (val == 0)
  return "0";
if ((val & 0x800) != 0)
  return fmt::format("-0x{:04X}", ((~val) & 0xfff) + 1);
return fmt::format("0x{:04X}", val);
```

TryTwo replied:
I removed the braces from the similar function above it too.

sepalani reviewed (commented):
Other than that, LGTM.

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
Ok, I think we're good.

sepalani reviewed (approved):
LGTM

AdmiralCurtiss commented:
It's a bit silly that you're reverse-engineering the register from the disassembled instruction string, but eh, seems to work well enough.
